.github/workflows/stacks-blockchain-tests.yml

```diff
@@ -1,117 +1,321 @@
+  cancel-in-progress: ${{ github.event_name == 'pull_request' }}
+
 jobs:
-  # Run full genesis test
+  ## Full genesis test with code coverage
   full-genesis:
     name: Full Genesis Test
     runs-on: ubuntu-latest
+    ## Run a maximum of 2 concurrent tests from the test matrix
+    strategy:
+      fail-fast: false
+      max-parallel: 2
+      matrix:
+        test-name:
+          - neon_integrations::bitcoind_integration_test
     steps:
+      ## Checkout the code
       - name: Checkout the latest code
         id: git_checkout
-        uses: actions/checkout@v3
-      - name: Single full genesis integration test
-        id: full_genesis_test
-        env:
-          DOCKER_BUILDKIT: 1
-        # Remove .dockerignore file so codecov has access to git info
+        uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
+
+      ## Install rust toolchain (llvm-tools-preview)
+      - name: Setup Rust Toolchain
+        id: setup_rust_toolchain
+        uses: actions-rust-lang/setup-rust-toolchain@f3c84ee10bf5a86e7a5d607d487bf17d57670965 # v1.5.0
+        with:
+          toolchain: stable
+          components: llvm-tools-preview
+          cache: false
+
+      ## Restore cargo cache data
+      - name: Restore Cargo Cache
+        id: restore_cargo_cache
+        uses: actions/cache/restore@88522ab9f39a2ea568f7027eddc7d8d8bc9d59c8 # v3.3.1
+        with:
+          fail-on-cache-miss: true
+          path: |
+            ~/.cargo/bin/
+            ~/.cargo/registry/index/
+            ~/.cargo/registry/cache/
+            ~/.cargo/git/db/
+          key: ${{ github.event.repository.name }}-${{ github.sha }}-cargo
+
+      ## Restore bitcoin cache data
+      - name: Restore Bitcoin Binary Cache
+        id: restore_bitcoin_cache
+        uses: actions/cache/restore@88522ab9f39a2ea568f7027eddc7d8d8bc9d59c8 # v3.3.1
+        with:
+          fail-on-cache-miss: true
+          path: /tmp/bitcoin
+          key: ${{ github.event.repository.name }}-bitcoin-binaries
+
+      ## Restore genesis nextest archive cache data
+      - name: Restore Genesis Archive Cache
+        id: restore_test-archive_cache
+        uses: actions/cache/restore@88522ab9f39a2ea568f7027eddc7d8d8bc9d59c8 # v3.3.1
+        with:
+          fail-on-cache-miss: true
+          path: /tmp/genesis_archive.tar.zst
+          key: ${{ github.event.repository.name }}-${{ github.sha }}-genesis-archive
+
+      ## Restore nextest target cache data
+      - name: Restore Target Cache
+        id: restore_target_cache
+        uses: actions/cache/restore@88522ab9f39a2ea568f7027eddc7d8d8bc9d59c8 # v3.3.1
+        with:
+          fail-on-cache-miss: true
+          path: ./target
+          key: ${{ github.event.repository.name }}-${{ github.sha }}-target
+
+      ## Symlink bitoind to $PATH
+      - name: Link Bitcoin Binary
+        id: bitcoin_binary
+        run: |
+          sudo ln -s /tmp/bitcoin/bin/bitcoind /bin/
+          ls -al /bin/bitcoind
```

Comment:
Seems like these blocks are duplicated in a few places.  Is there an opportunity to DRY this up?

Reply:
with the current workflow, i don't think it's possible to DRY this. since the bitcoin binary needs to be present for multiple jobs in the workflow. 

I have an idea on how to DRY this, but i'm not sure if:
a) it'll even work
b) it won't make things messier

i can create a test for this and see what happens. 

as to your and @kantai ask about workflows that have run - i think i can get a few execs in the the stacks-blockchain repo so you an see how it works in practice vs my personal fork where i've been testing these changes. 

in light of https://github.com/stacks-network/stacks-blockchain/actions/runs/6511171596/job/17688907514?pr=3950 as well, i think i'm going to add some logic to perform a consistent cleanup of the VM across several workflows (vs targeting them when we hit an error like @xoloki noticed in the linked action)